pyicloud/services/drive.py

```diff
@@ -17,11 +20,12 @@ def __init__(self, service_root, document_root, session, params):
 
     def _get_token_from_cookie(self):
         for cookie in self.session.cookies:
-            if cookie.name == "X-APPLE-WEBAUTH-TOKEN":
+            if cookie.name == "X-APPLE-WEBAUTH-VALIDATE":
+                print(cookie.value)
                 match = search(r"\bt=([^:]+)", cookie.value)
-                if not match:
+                if match is None:
                     raise Exception("Can't extract token from %r" % cookie.value)
-                self.params.update({"token": match.group(1)})
+                return {"token": match.group(1)}
```

Comment:
I think we still can use `self.params`, and call `_get_token_from_cookie` at init.

So it won't search on the cookie everytime we want to use `_get_upload_contentws_url`, but only once.

In `_get_upload_contentws_url` use `self.params`

Reply:
That was the original code and I thought about it but I do not think it is the right approach.

Adding it to self.params means that subsequent GET/POST etc. requests will have the token parameter appended to the URL as a query parameter.
The way icloud handles it seems to be that the token parameter is _only_ added on this specific call.
I suspect this might be because there's a chance that the redirect might go to somewhere outside the .icloud.com domain, which means no Cookie would be sent and thus the token is needed... But this is pure conjecture on my part.

As an aside: It seems that a token of "NONE" works right now as well, because we're inside the icloud.com domain which means cookies are sent... But I thought it'd be prudent to emulate the way the browser is handling things for now...